pylint/plugins/hass_enforce_type_hints.py

```diff
@@ -1697,12 +1788,15 @@ def visit_module(self, node: nodes.Module) -> None:
         ):
             self._class_matchers.extend(property_matches)
 
+        self._class_matchers.reverse()
+
     def visit_classdef(self, node: nodes.ClassDef) -> None:
         """Called when a ClassDef node is visited."""
         ancestor: nodes.ClassDef
         checked_class_methods: set[str] = set()
-        for ancestor in node.ancestors():
-            for class_matches in self._class_matchers:
+        ancestors = list(node.ancestors())  # cache result for inside loop
+        for class_matches in self._class_matchers:
+            for ancestor in ancestors:
```

Comment:
@cdce8p in a previous PR you requested that we invert the two loops to avoid multiple calls to `node.ancestors()`
I have found an issue with mixed inheritance where we would sometimes hit `Entity` before hiting `NumberEntity`.

I have added a test case for this, and caching the results together with inverting the class matchers has fixed this.

Reply:
The fix looks good! I appreciate the test case. It helped me better understand the issue.